doc/architectural/background-concepts.md

```diff
@@ -1,32 +1,810 @@
+expression: i.1.2 = i.1.1 + 1
+```
+This representation now allows us to state facts such as
+''`i` is increasing'' by writing `i.1.1 < i.1.2`.
+
+At this point, we run into a complication. Consider the following piece
+of code for illustration:
+```C
+// Given some integers a, b
+int x = a;
+if (a < b)
+  x = b;
+return x;
+```
+The corresponding control flow graph looks like this:
+
+\dot "CFG for maximum function"
+digraph ast {
+  node [shape=rectangle]
+  first [label="declare x\nx = a\na < b"]
+  second [label="x = b"]
+  third [label="return x"]
+  first -> second [label="true"]
+  first -> third [label="false"]
+  second -> third
+}
+\enddot
+
+When we try to transform to SSA, we get:
+\dot "CFG for maximum function - SSA, attempt"
+digraph ast {
+  node [shape=rectangle]
+  first [label="\nx.1 = a\na < b"]
+  second [label="x.2 = b"]
+  third [label="return ???"]
+  first -> second [label="true"]
+  first -> third [label="false"]
+  second -> third
+}
+\enddot
+Depending on which path the execution takes, we have to return either
+`x.1` or `x.2`! The way to make this work is to introduce a function
+&Phi; that selects the right instance of the variable; in the
+example, we would have
+\dot "CFG for maximum function - SSA using &Phi;"
+digraph ast {
+  node [shape=rectangle]
+  first [label="\nx.1 = a\na < b"]
+  second [label="x.2 = b"]
+  third [label="return &Phi;(x.1,x.2)"]
+  first -> second [label="true"]
+  first -> third [label="false"]
+  second -> third
+}
+\enddot
+
+In CBMC, we use a slightly different solution. For each branch, we keep
+track of which side of the branch was taken, and implement &Phi; using
+this information. There are also some differences in how loops are
+handled (finite unrolling in CBMC, versus a &Phi;-based approach in
+compilers); the CBMC approach will be discussed in a later chapter.
+
+For the time being, let us come back to `factorial`. We can now give
+an SSA using &Phi; functions:
+
+\dot "Control flow graph in SSA for factorial"
+digraph cfg {
+  node [shape=rectangle]
+  bb1 [label="unsigned long fac.1.1 = 1\nunsigned i.1.1 = 1",peripheries=2]
+  bb2 [label="i.1.2 = &Phi;(i.1.1, i.1.3)\nfac.1.2 = &Phi;(i.1.1, i.1.3)\ni.1.2 <= n.1.1"]
+  bb3 [label="fac.1.3 = fac.1.2 * i.1.2\ni.1.3 = i.1.2 + 1"]
+  
+  bb4 [label="Return fac.1.2",style=filled,fillcolor=gray80]
+  {bb1, bb3} -> bb2
+  bb2 -> bb3 [label="true"]
+  bb2 -> bb4 [label="false"]
+}
+\enddot
+
+TODO: Point to some good discussion of SSA
+
+The SSA is an extremely helpful representation when one
+wishes to perform model checking on the program (see next section),
+since it is much easier to extract the logic formulas used in this
+technique from an SSA compared to a CFG (or, even worse, an AST). That
+being said, CBMC takes a different route, opting to convert to
+intermediate representation known as GOTO programs instead.
+
+\subsection Trace_section Trace
+
+TODO: I am not quite sure what is supposed to go here.
```

Comment:
I think this was intended to describe how a (counterexample) trace is represented.